### ISSUE TITLE
About training on fine + coarse data set

### ISSUE THREAD

**Issue by xiongzhanblake** [holyseven/PSPNet-TF-Reproduce#4]
Hi, thanks for your nice work！ Recently, I am following this pspnet work.
I am curious about how you trained on the fine + coarse data set. Here is my idea:
1. split training into 2 steps: train on the coarse or fine data set first and on the another then.
2. mix fine and coarse data set, and train once only.

**Comment by holyseven** [comment]
I obtained the results following the process below, (proposed by authors of PSPNet): 
- train on extra+train (20000+2975 images), 
- then on train set (2975), 
- test on val set (500) or test on test set (1525).

I tried `training once`, not very good. No idea how to do it properly and why.

**Comment by xiongzhanblake** [comment]
Thanks for your response : )
So the results showed in [Numerical Results](https://github.com/holyseven/PSPNet-TF-Reproduce#numerical-results) are obtained from the above process. Quite impressive results! The poor performance of once training, I suppose， might result from two different sets of annotations, i.e. gt_coarse and gt_fine. 
Hereafter, I came back to original paper and guess the process may be:
1. train on extra+train with coarse labels.
2. then on train set with fine labels.
3. test on val set or test set.

**Comment by xiongzhanblake** [comment]
One more question:
Did you transfer all ground truth files '*labelIds.png' into '*labelTrainIds.png' following this [preparation](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdInstanceImgs.py) and after then put all transferred png files into your gt folder? 
Thanks!

**Comment by holyseven** [comment]
I used only 19 classes for training (i.e. TrainIds) and test. The preparation followed the [id definition](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py) but was not done by this [script](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdInstanceImgs.py).

As for training with coarsely labeled data, whether the first training stage should involve the fine labeled data, I don't know ... But both the results I obtained and those from the original paper are obtained from the process I mentioned before (verified with the authors). Note that there is no strong argument (at least I haven't found one) that this is proper, but this is probably the best practice.

Two more details when training with extra_coarse data:
1. In the original paper of PSPNet, the learning rate for new layers is set to 10 times larger than the existing layers. I named this learning policy `train_like_in_paper`, which can be found in the definition of `parser` in train.py.
2. In the original work, the first training stage has a base learning rate 0.01 with 150K iterations; in the second stage, a base learning rate 0.001 with 90K iterations.

However, I trained with `train_like_in_paper = 0` in both stages, and the second stage was trained with a base learning 0.01 and 50K iterations.

**Comment by xiongzhanblake** [comment]
I believe that in the first stage gt_fine should be excluded because this stage works as a coarsely / weakly training. Then finely train with gt_fine in the second stage.

**Comment by lxtGH** [comment]
Hi！ @holyseven 
 Thanks for your code, what kind of gpu you use? what is its memory ? (I guess 12GB above), I use the code with then same setting in your README file, since my gpu memory limits(11GB), I 8 GPU so one image per GPU, the model doesn't converge. the loss is still about 1.5 after 50000 iteration

**Comment by holyseven** [comment]
@lxtGH On Cityscapes? Have you done the preparation as we discussed here? Would you share your training script and some of your log file?

```
{'batch_size': 2, 'blur': 1, 'bn_frozen': 0, 'color_switch': 0, 'data_type': 32, 'database': 'Cityscapes', 'ema_decay': 0.9, 'epsilon': 1e-05, 'eval_only':
 0, 'fine_tune_filename': '../z_pretrained_weights/resnet_v1_101.ckpt', 'fisher_epsilon': 0, 'fisher_filename': './fisher_exp.npy', 'fix_blocks': 0, 'gpu_n
um': 4, 'has_aux_loss': 1, 'initializer': 'he', 'log_dir': 'only-resnet', 'loss_type': 'normal', 'lr_step': None, 'lrn_rate': 0.01, 'mirror': 1, 'momentum'
: 0.9, 'network': 'pspnet', 'new_layer_names': None, 'norm_only': 0, 'optimizer': 'mom', 'poly_lr': 1, 'random_rotate': 0, 'random_scale': 1, 'resize_image
s_method': 'bilinear', 'resume_step': None, 'save_first_iteration': 0, 'scale_max': 2.0, 'scale_min': 0.5, 'server': 2, 'snapshot': 25000, 'step_size': 0.1
, 'structure_in_paper': 0, 'subsets_for_training': 'train', 'test_batch_size': 5, 'test_image_size': 864, 'test_max_iter': None, 'train_conv2dt': 0, 'train
_image_size': 864, 'train_like_in_paper': 0, 'train_max_iter': 50000, 'weight_decay_mode': 1, 'weight_decay_rate': 0.0001, 'weight_decay_rate2': 0.0001}
================ using tf.float32 =====================
['train']
Database has 2975 images.
applying random mirror ...
applying random scale [0.500000, 2.000000]...
================ Database Info ================
================== Resnet structure =======================
new layer names:  ['block4/unit_4', 'block3/unit_24', 'logits', 'psp']
num_residual_units:  [3, 4, 23, 3]
channels in each block:  [256, 512, 1024, 2048]
stride in each block:  [1, 2, 1, 1]
rates in each atrous convolution:  [1, 1, 2, 4]
================== constructing network ====================
```

And feel free to open a new issue if you think this is a different subject from here.

**Comment by lxtGH** [comment]
Hi ! @holyseven Thank you for your reply.
I found my problem, I used the wrong labels, I check your code, I see you use **labeltrainids**
so the model dosen't converge after 50000 iteration.
maybe you can add following code in reader_segmentation.py , function _read_cityscapes_image_label_list for others use.
`labels_filename_proto = data_dir + '/gtFine/' + data_sub + '/*/*labelTrainIds.png'`
`assert len(labels) == len(images), "train images and labels must equal"` 
Now I will retrain the model.
Thanks again for your reply !

**Comment by holyseven** [comment]
@lxtGH Thanks! Done!

**Comment by muaz-git** [comment]
> I used only 19 classes for training (i.e. TrainIds) and test. The preparation followed the [id definition](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py) but was not done by this [script](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdInstanceImgs.py).
> 
> As for training with coarsely labeled data, whether the first training stage should involve the fine labeled data, I don't know ... But both the results I obtained and those from the original paper are obtained from the process I mentioned before (verified with the authors). Note that there is no strong argument (at least I haven't found one) that this is proper, but this is probably the best practice.
> 
> Two more details when training with extra_coarse data:
> 
> 1. In the original paper of PSPNet, the learning rate for new layers is set to 10 times larger than the existing layers. I named this learning policy `train_like_in_paper`, which can be found in the definition of `parser` in train.py.
> 2. In the original work, the first training stage has a base learning rate 0.01 with 150K iterations; in the second stage, a base learning rate 0.001 with 90K iterations.
> 
> However, I trained with `train_like_in_paper = 0` in both stages, and the second stage was trained with a base learning 0.01 and 50K iterations.

Thanks for this amazing repo. I am currently trying to follow this strategy for training cityscapes' dataset. However, for second stage, I am not quite sure if there is a process of "resuming" training  currently, or maybe I am missing something. 

Also, we do not need the param `fine_tune_filename` (with `resnet_v1_101` or `stage 1` weights) during second stage for calculation of `wd` right? 

Would be great to have your comments on it.

**Comment by holyseven** [comment]
> for second stage, I am not quite sure if there is a process of "resuming" training currently, or maybe I am missing something.

The moments and batch normalization global statistics are not used for the second-stage training, but all trainable parameters are.
When `new_layer_names` is set to '' or 'whatever' (default is `None`), then all trainable parameters will be loaded from the `fine_tune_filename`.

> we do not need the param fine_tune_filename (with resnet_v1_101 or stage 1 weights) during stage for calculation of wd right?

For normal weight decay (L2), no; For L2-SP, we need the `fine_tune_filename` as reference for regularization.

**Comment by muaz-git** [comment]
Thank you for your quick response. 
So I am trying to achieve the same results on `cityscapes` using PSPNet as you did. I have followed these steps. 

- Converted labelIds to trainIds for `train`, `train_extra` and `val` splits.
- Training happens in two stages, first stage uses following arguments:
`{'batch_size': 2, 'color_switch': 0, 'coloring': 1, 'consider_dilated': 0, 'database': 'Cityscapes', 'fine_tune_filename': './z_pretrained_weights/resnet_v1_101.ckpt', 'initializer': 'he', 'lr_step': None, 'lrn_rate': 0.01, 'mirror': 1, 'momentum': 0.9, 'ms': 0, 'network': 'resnet_v1_101', 'new_layer_names': None, 'optimizer': 'mom', 'poly_lr': 1, 'prefix': 'PSP', 'random_blur': 1, 'random_mirror': 1, 'random_rotate': 0, 'random_scale': 1, 'reader_method': 'queue', 'resume_step': None, 'save_first_iteration': 0, 'scale_max': 2.0, 'scale_min': 0.5, 'snapshot': 15000, 'step_size': 0.1, 'subsets_for_training': 'train,train_extra', 'test_image_size': 864, 'test_subset': 'val', 'three_convs_beginning': 0, 'train_image_size': 864, 'train_like_in_caffe': 0, 'train_max_iter': 150000, 'visible_gpus': '0,1', 'weight_decay_mode': 1, 'weight_decay_rate': 0.0001, 'weight_decay_rate2': 0.0001}`, in summary trained on `renset_v1_101` pretrained (on ImageNet) for 150k steps with `lrn_rate`:`0.01`. And most importantly using both `train` and `train_extra` splits (only fine annotations for train split, and coarse annotations only for train_extra split). After 150k steps mIOU of `~69` is achieved on validation set.
- During stage 2, all arguments remain the same except for the following:
`{'fine_tune_filename': <Path to CHECKPOINT STAGE 1>,  'subsets_for_training': 'train', 'train_max_iter': 50000}`. In summary training for only 50k steps with train split only over stage 1 pretrained network. And after second stage could only achieve mIOU of `~77` on validation set.

Can you please comment over how can I manage to achieve mIOU of `80`?

**Comment by holyseven** [comment]
`batch_size` * num of `visible_gpus` should be larger than 8 or 16.
